### DISCUSSION TITLE
How do I get Google Authentication to work with Flutter Web?

### DISCUSSION THREAD

**Discussion by coryfail** [pocketbase/pocketbase#4599] [Q&A] [answered]
Google Auth works just fine using the way it is shown in docs on iOS, however, for web the `launchUrl()` function wasn't working correctly. It would open the Google Auth login, but never redirect away from the login page. I added some prints and determined `launchUrl()` was the issue.

For testing, I replaced it with ` html.window.location.href = url.toString();`

 ```dart
Future<void> loginProvider(String providerName) async {
    try {
      final authData = await pb.collection('users').authWithOAuth2(providerName,
          (url) async {
        html.window.location.href = url.toString();
      });
      await _secureStorageService.saveToken(authData.token);

      isUserAuthenticated.value = true;
      Get.offAllNamed('/dashboard');
    } catch (e) {
      isUserAuthenticated.value = false;
      print(e);
    }
  }
```

This now gets me further but now I get the attached error:

<img width="1085" alt="Screenshot 2024-03-20 at 9 11 39 AM" src="https://github.com/pocketbase/pocketbase/assets/7023376/2f9257ea-ba1f-4713-87b1-49f651142802">

I even tried using ` await launchUrl(url, webOnlyWindowName: '_blank');` with the same results.

Here is my Google Auth config:

<img width="515" alt="Screenshot 2024-03-20 at 9 20 08 AM" src="https://github.com/pocketbase/pocketbase/assets/7023376/0d47d056-8538-45da-8a2a-3eea6668a51f">

It's interesting that this isn't working for web but it does mobile since it uses the web google auth configuration for all of it.

**Comment by ganigeorgiev** [comment, marked as answer]
The "all-in-one" OAuth2 flow needs to establish a realtime connection with the PocketBase server in order to receive the OAuth2 result without a deeplink or extra redirect, but as mentioned in the [Limitations section](https://github.com/pocketbase/dart-sdk?tab=readme-ov-file#limitations) of the SDK, `dart-lang/http` doesn't support streaming responses on the web out of the box and a custom platform specific client must be provided.
The readme shows how for example you can load the [`fetch_client` package](https://pub.dev/packages/fetch_client):

```dart
import 'package:pocketbase/pocketbase.dart';
import 'package:fetch_client/fetch_client.dart';
import 'package:flutter/foundation.dart' show kIsWeb;

final pb = PocketBase(
  'http://127.0.0.1:8090',
  httpClientFactory: kIsWeb ? () => FetchClient(mode: RequestMode.cors) : null, // load fetch_client for web, otherwise fallback to the default dart-lang/http
);
```

**Reply by coryfail** [reply]
I'll give that a go. I did see you mentioned this to me in our last discussion, but when I read this documentation it wasn't clear this specific limitation was part of the "all-in-one" OAuth2 flow issue I am having. May I suggest that we slightly re-word it so others have a clear understanding what the Limitation effects?

**Reply by ganigeorgiev** [reply]
_I've updated the readme to mention the "all-in-one" OAuth2 flow as part of the streamed responses web limitation._

**Reply by coryfail** [reply]
@ganigeorgiev FYI your example in your limitations is slightly incorrect. This will not build for both web and mobile since it contains web only packages.

I made an example of how I fixed this issue here:
https://github.com/coffeebean-dev/flutter-getx-pocketbase?tab=readme-ov-file#work-around-for-limitations
